File: LeetCode/patterns/prefix_sum_pattern/contiguous_array_leetcode_525.py
"""
Given a binary array nums, return the maximum length of a contiguous subarray with an equal number of 0 and 1.

 

Example 1:

Input: nums = [0,1]
Output: 2
Explanation: [0, 1] is the longest contiguous subarray with an equal number of 0 and 1.
Example 2:

Input: nums = [0,1,0]
Output: 2
Explanation: [0, 1] (or [1, 0]) is a longest contiguous subarray with equal number of 0 and 1.
 

Constraints:

1 <= nums.length <= 10^5
nums[i] is either 0 or 1.
"""

from typing import List

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        # Ref Video: https://www.youtube.com/watch?v=agB1LyObUNE
        # 111 000
        # How to know to shrink window or increase?
        # Have to split, thus recursive and worse than O(n^2)

        # Not every result starts at beginning, e.g. 111 00

        # is it possible to know exactly where the first 1 was, and verify that after it
        # count[1] == count[0]

        # map each index to pair (count[0], count[1])
        # Use diff between counts for O(1) lookup of what we need
        
        zero, one = 0, 0
        res = 0
        diff_index = {} # diff --> index   [diff=count[1]-count[0]]
        
        for i, n in enumerate(nums):
            if n == 0:
                zero += 1
            else:
                one += 1
        
            if one - zero not in diff_index:
                diff_index[one - zero] = i
            
            if one == zero:
                res = one + zero # This is definitely the longest subarray at this point
            else:
                idx = diff_index[one - zero]
                res = max(res, i - idx)
        
        return res

File: LeetCode/patterns/prefix_sum_pattern/contiguous_array_leetcode_525.java
class Solution {
    public int findMaxLength(int[] nums) {
        if (nums.length == 0) return 0;
        Map<Integer, Integer> count_dict = new HashMap<>();
        int res = 0;
        int count = 0;

        count_dict.put(0, -1);
        
        for (int i = 0; i < nums.length; i++){
            int num = nums[i];
            if (num == 1) {
                count = count + 1;
            } else {
                count = count - 1;
            }

            if (!count_dict.containsKey(count)) {
                count_dict.put(count, i);
            } else {
                res = Math.max(res, i - count_dict.get(count));
            }
        }
        return res;
    }
}
